### PR TITLE
mpiserializer: handle all vectors

### PR DESCRIPTION
**Pull request by akva2**
Handle all vectors in eclmpiserializer and remove support in MPIPacker.

This contains some more changes, in particular we need to move the variadic broadcast support into the serializer since that previously relied on the vector support in the MPIPacker.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: ebos/eclmpiserializer.hh [hunk 1]
```diff
@@ -95,12 +95,16 @@ public:
...
             optional(data);
+        } else if constexpr (is_vector<T>::value) {
+            vector(const_cast<T&>(data));
         } else if constexpr (is_map<T>::value) {
             map(const_cast<T&>(data));
         } else if constexpr (is_array<T>::value) {
             array(const_cast<T&>(data));
         } else if constexpr (is_set<T>::value) {
             set(const_cast<T&>(data));
+        } else if constexpr (has_serializeOp<detail::remove_cvr_t<T>>::value) {
```
Comment (blattms) [inline ebos/eclmpiserializer.hh:106]: Would have thought that this removes  `volatile` but it does not...

File: ebos/eclmpiserializer.hh [hunk 2]
```diff
@@ -391,28 +413,66 @@ public:
             if (m_packSize == std::numeric_limits<size_t>::max()) {
                 throw std::runtime_error("Error detected in parallel serialization");
             }
+
             m_buffer.resize(m_packSize);
             m_comm.broadcast(m_buffer.data(), m_packSize, 0);
```
Comment (atgeirr) [inline ebos/eclmpiserializer.hh:418]: Er 0 her riktig?

Comment (akva2) [reply ebos/eclmpiserializer.hh:418]: nope, fixed.

**Review by blattms — APPROVED** [review]
One a side note:

I think we might loose something (speed memory) due this refactoring for contiguous arrays/vectors of pods. Those should actually be sized/packed/unpacked in one go by specifying the number of entries instead of using a loop over the  entries and make the operation on each entry. Hope hat can be changed.

File: ebos/eclmpiserializer.hh (hunk 1 above)
Comment (akva2) [reply ebos/eclmpiserializer.hh:106]: indeed, i forgot to remove volatile.

**Comment by akva2** [comment]
as vectors of pod; maybe, but this is not a regression. it was never actually utilized in the old MPIPacker support either. it was also a loop over the entries.

**Comment by akva2** [comment]
jenkins build this please

File: ebos/eclmpiserializer.hh [hunk 3]
```diff
@@ -64,7 +64,7 @@ decltype(auto) make_variant(std::size_t index)
 }
 
 template<class T>
-using remove_cvr_t = std::remove_const_t<std::remove_reference_t<T>>;
+using remove_cvr_t = std::remove_volatile_t<std::remove_const_t<std::remove_reference_t<T>>>;
```
Comment (bska) [inline ebos/eclmpiserializer.hh:67]: For what it's worth, `std::remove_cv_t<>` removes both `const` and `volatile`.

Comment (akva2) [reply ebos/eclmpiserializer.hh]: simplified.

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
merging with approval.

**Comment by bska** [comment]
Did this [break](https://ci.opm-project.org/job/opm-simulators/1363/) the sequential build?

**Comment by akva2** [comment]
it did. fix in https://github.com/OPM/opm-simulators/pull/4105

**Pull request merged**